### ISSUE TITLE
Add a requriment validator

### ISSUE THREAD

**Issue by FFY00** [pypa/packaging#317]
As discussed on IRC, I propose adding a requirement validator to `packaging`. There's is a real need for it, I need it for at least some of my projects.

It would use `importlib.metadata` to find the metadata.

**Comment by pradyunsg** [comment]
Are you asking for https://github.com/pypa/packaging/issues/147?

**Comment by FFY00** [comment]
No, sorry. I wasn't clear enough. I am asking for a way to validate if the requirements are met. This can be used for eg. in a python-build to validate the build dependencies are met, or in a wheel installer, to validate the make sure the dependencies are available, and compatible, before installing the wheel.

**Comment by FFY00** [comment]
I have a minimal implementation, but it does not validate the dependencies of extras. It would be good to have a strong implementation here, since it turns out it is not trivial.

https://github.com/FFY00/python-build/blob/master/build/__init__.py#L39

**Comment by pradyunsg** [comment]
I see, something like `pip check` but independent of pip?

**Comment by FFY00** [comment]
Yes.

**Comment by brettcannon** [comment]
One thing to consider about this proposal is I think this would be the first bit of functionality that actually queries installed packages. Everything else in 'packaging' is mostly parsing. `packaging.tags` is the only thing that queries the live interpreter running, but we don't have anything that goes into the environment to accomplish something.

Now I'm not saying expanding out is bad, I just want to be upfront it does open up the scope of this project a bit (and bring in a dependency on importlib-metadata).

**Comment by FFY00** [comment]
I agree, we should consider if we should use `importlib.metadata` or something else.

I would say the `importlib-metadata` dependency is just a temporary nuisance, `importlib.metdata` is available from Python 3.8 and up. Given time, older versions a compatible Python version will start to get adopted. If the `importlib-metadata` dependency is still a problem, we can declare an extra that pulls it.

**Comment by pradyunsg** [comment]
FWIW, I'm certainly not 100% sure if packaging is the correct place to put it. It might be that importlib should grow a helper function to do this or something.

**Comment by FFY00** [comment]
But it would need to use the functionality provided by `packaging`. As `importlib` is part of the standard library, this is not possible.

I agree `packaging` might not be the most optimal place, but given the context, I believe it is where it makes most sense. And the functionality proposed falls somewhat in scope of the project.

Do you have any place in mind other than `importlib`?

**Comment by brettcannon** [comment]
I agree that I don't think importlib is a good fit for this.

Would this big a good fit for pradyunsg/installer?

**Comment by FFY00** [comment]
I think it makes more sense to have it here. This library aims to provide "Core utilities for Python packages", I think verifying if the requirements are met is one of those workflows. pradyunsg/installer provides a library with the install logic, I do not really believe this is install logic, IMO "packaging logic" better describes it.

**Comment by layday** [comment]
Regardless of what happens here I think it would be useful to expose individual markers in `markers.Marker` so we can e.g. filter requirements based on their marker value for extras.  A related issue is that `Marker.evaluate` will raise if its own extra is missing from the environment, so requirements cannot be evaluated if they contain an extra.

**Comment by FFY00** [comment]
We have a working implementation in pypa/build. How do you we want to proceed?

**Comment by FFY00** [comment]
Sorry to bother, but we have working code for this. Would a PR be welcome?

**Comment by brettcannon** [comment]
I'm still on the fence for this since it just feel like it quite fits into 'packaging' (don't ask me why; gut feel), but then I don't know where it would belong in PyPA either.

**Comment by FFY00** [comment]
Yeah, and I agree that pypa/packaging might not be the best place to put it, but I think it is close enough to be relevant and from all other projects I think pypa/packaging is what makes the most sense.

I would really like to figure this out, otherwise it will be indefinitely stalled :confused:

This simply integrates the pypa/packaging functionality with stdlib's importlib.metadata.

It feels silly to create a new package just for this, but that could be done if we don't arrive to any consensus. I mean, maybe we could bundle a few package introspection routines, but I struggle to make a strong case for this new package.

**Comment by layday** [comment]
I suppose we could use a callback in place of `importlib.metadata` if that makes it more palatable for inclusion in `packaging`.  I've adapted the implementation from `build` with this in mind: https://github.com/layday/build/blob/prospective-packaging-implementation/check_dependencies.py.
